docs/GDPs/gdp-0002.rst

```diff
@@ -0,0 +1,346 @@
+
+The edge objects are wrappers that provide ranges to a `Box` class:
+
+.. code-block:: python
+
+    with region(Box(slice(0+grid.global_istart, 1+grid.global_istart), slice(0, N))), interval(...):
+
+When GT4Py asks for the iteration space from the Box, it sees different slices on each process using the local iteration space indices:
+
+.. code-block:: python
+
+    proc 0:    ((slice(0, 1), slice(0, N)),)
+    proc 1:    (,)
+    proc 2:    ((slice(n-1, n), slice(0, N)),)
+
+The result of the evaluation is inserted directly into code, so GT4Py and its backend never deal with partitioning.
+Translation to the local indices is done within the Box object specialization from the application.
+
+Note that if this example had only one process, the local and global iteration spaces coincide, so the slices set simply get passed on to GT4Py unmodified.
+
+
+Compilation-Time Evaluation
++++++++++++++++++++++++++++
+
+The expression `expr` inside `region(expr)` is evaluated at compile-time, and the resulting index ranges are stored in the library module created.
+Importantly, this means that GT4Py does not need to create objects to mirror these expression at call-time, as they may call partitioners or other application-specific code.
+
+
+Box Objects
++++++++++++
+
+The `Box` slice data could be coverted into a `local_origin` and a `local_domain`, to mirror the existing implementation and stay compatible with GridTools boundary conditions.
+If the `local_domain` is of size `0` in any dimension, the block will not execute.
+This information could be utilized by the backend to optimize not computing unnecessary stencils.
+Note that `region(*args)` so the same sub-stencil can execute on any number of boxes passed, as was shown in the motivating example.
+
+Application developers will create their own specializations of Box for their particular uses.
+For example, since our use case has edges and corners, we may have specific `Corner` specializations as well.
+
+When using `region()`, the stencil caller should avoid using `origin` and `domain` as this would change the iteration space and would be very error prone.
+
+Halo Regions
+++++++++++++
+
+It is important for our application that halo regions can be accessed by the stencils.
+This falls out naturally from the `region()` by passing negative integers to the `Edge` boxes.
+For example, `WestEdge(-2,0)` will access two lines of halo points.
+
+
+Corner Cases and Warnings
++++++++++++++++++++++++++
+
+It may be the case that a part of the iteration space is accessed twice by different `with` blocks.
+These should ideally be scheduled and run in the order they appear, but as a first implementation GT4Py could throw an error in this case.
+An option should be added to determine if this should be an error.
```

Comment:
We want it to not throw an error eventually, I think, because there are cases where the code first goes over the entire region and then it corrects corners and edges.

Reply:
Already removed. Thanks.